File: reference/flutter-src/lib/data/sync/sync_service.dart
import '../../domain/models/detection.dart';

/// ┌──────────────────────────────────────────────────────────────────────┐
/// │  BACKEND SEAM — SyncService (optional, Firebase)                      │
/// │                                                                        │
/// │  The app is fully functional offline; sync is additive. This interface │
/// │  is the only place the app talks to a backend. The default binding is  │
/// │  NoopSyncService (does nothing), so nothing depends on Firebase.       │
/// │                                                                        │
/// │  To enable cross-device sync + aggregate maps + model updates:         │
/// │   1. Add firebase_core + cloud_firestore, run flutterfire configure.   │
/// │   2. Implement FirebaseSyncService against this interface (push local  │
/// │      detections, pull aggregates).                                     │
/// │   3. Override the provider in main.dart. No UI/domain changes.         │
/// └──────────────────────────────────────────────────────────────────────┘
abstract interface class SyncService {
  bool get isEnabled;

  /// Push newly-saved detections upstream (best-effort). Never blocks the UI.
  Future<void> pushDetection(Detection detection);

  /// Pull remote aggregates for the map. Returns empty when disabled.
  Future<void> pullAggregates();
}

/// Default no-op binding — keeps the app 100% offline unless Firebase is wired.
class NoopSyncService implements SyncService {
  const NoopSyncService();

  @override
  bool get isEnabled => false;

  @override
  Future<void> pushDetection(Detection detection) async {}

  @override
  Future<void> pullAggregates() async {}
}

File: app/src/main/kotlin/com/mozzid/domain/sync/SyncService.kt
package com.mozzid.domain.sync

import com.mozzid.domain.model.Detection

/**
 * ┌────────────────────────────────────────────────────────────────────────┐
 * │  BACKEND SEAM — SyncService (optional, Firebase)                        │
 * │                                                                          │
 * │  The app is fully functional offline; sync is additive. This interface   │
 * │  is the only place the app talks to a backend. The default binding is     │
 * │  NoopSyncService (does nothing), so nothing depends on Firebase.          │
 * │                                                                          │
 * │  To enable cross-device sync + aggregate maps + model updates:            │
 * │   1. Add firebase-bom + firestore + auth, drop in google-services.json.   │
 * │   2. Implement FirebaseSyncService against this interface.                │
 * │   3. Swap the binding in Bootstrap. No UI/domain changes.                 │
 * └────────────────────────────────────────────────────────────────────────┘
 */
interface SyncService {
    val isEnabled: Boolean

    /** Push a newly-saved detection upstream (best-effort). Never blocks the UI. */
    suspend fun pushDetection(detection: Detection)

    /** Pull remote aggregates for the map. No-op when disabled. */
    suspend fun pullAggregates()
}
